metricbeat/module/http/json/json.go

```diff
@@ -0,0 +1,138 @@
+package json
+
+import (
+	"encoding/json"
+	"io/ioutil"
+	"net/http"
+	"strings"
+
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/logp"
+	"github.com/elastic/beats/metricbeat/helper"
+	"github.com/elastic/beats/metricbeat/mb"
+	"github.com/elastic/beats/metricbeat/mb/parse"
+)
+
+// init registers the MetricSet with the central registry.
+// The New method will be called after the setup of the module and before starting to fetch data
+func init() {
+	if err := mb.Registry.AddMetricSet("http", "json", New, hostParser); err != nil {
+		panic(err)
+	}
+}
+
+const (
+	// defaultScheme is the default scheme to use when it is not specified in the host config.
+	defaultScheme = "http"
+
+	// defaultPath is the dto use when it is not specified in the host config.
+	defaultPath = ""
+)
+
+var (
+	hostParser = parse.URLHostParserBuilder{
+		DefaultScheme: defaultScheme,
+		PathConfigKey: "path",
+		DefaultPath:   defaultPath,
+	}.Build()
+)
+
+// MetricSet type defines all fields of the MetricSet
+// As a minimum it must inherit the mb.BaseMetricSet fields, but can be extended with
+// additional entries. These variables can be used to persist data or configuration between
+// multiple fetch calls.
+type MetricSet struct {
+	mb.BaseMetricSet
+	namespace string
+	http      *helper.HTTP
+	method    string
+	body      string
+}
+
+// New create a new instance of the MetricSet
+// Part of new is also setting up the configuration by processing additional
+// configuration entries if needed.
+func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
+
+	logp.Warn("The http json metricset is in beta.")
+
+	config := struct {
+		Namespace string `config:"namespace" validate:"required"`
+		Method    string `config:"method"`
+		Body      string `config:"body"`
+	}{}
+
+	if err := base.Module().UnpackConfig(&config); err != nil {
+		return nil, err
+	}
+
+	http := helper.NewHTTP(base)
+	http.SetMethod(config.Method)
+	http.SetBody([]byte(config.Body))
+
+	return &MetricSet{
+		BaseMetricSet: base,
+		namespace:     config.Namespace,
+		method:        config.Method,
+		body:          config.Body,
+		http:          http,
+	}, nil
+}
+
+// Fetch methods implements the data gathering and data conversion to the right format
+// It returns the event which is then forward to the output. In case of an error, a
+// descriptive error must be returned.
+func (m *MetricSet) Fetch() (common.MapStr, error) {
+
+	response, err := m.http.FetchResponse()
+	if err != nil {
+		return nil, err
+	}
+	defer response.Body.Close()
+
+	var jsonBody map[string]interface{}
+
+	body, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	err = json.Unmarshal(body, &jsonBody)
+	if err != nil {
+		return nil, err
+	}
+
+	event := common.MapStr{
```

Comment:
I would change this code as following. Note: code will not work directly probably because of some type conversion that is needed.

```
event := jsonBody
event["_namespace"] = m.namespace

if m.config.request.enabled {
	event[mb.ModuleData]["request"] = common.MapStr{
		"headers": m.getHeaders(response.Request.Header),
		"method":  response.Request.Method,
		"body":    m.body,
	}
}

if m.config.response.enabled {
	event[mb.ModuleData]["response"] = common.MapStr{
		"headers": m.getHeaders(response.Request.Header),
		"method":  response.Request.Method,
		"body":    m.body,
	}
}
```

The thinking behind this is as following:

* Httpbeat json is used either for self build http endpoints or for endpoints like spring which are not supported yet by Metricbeat. In these cases the json body contains all the relevant data. So in most cases the response header details is not too relevant if it is a 200. 
* The json document should not be nested under `body` as we already have the namespace to put it under its separate namespace.
* There are definitively use cases where request and response can become interesting, so we should have a config option to enable this. Be aware that these use cases will also overlap potentially with heartbeat and heartbeat could be the better tool for this use case.